File: moshi/moshi/inference_utils/turn_manager.py
# Copyright (c) Kyutai, all rights reserved.
# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.

"""Handling speaker transitions, text buffering, and context management for text display and RAG reference text generation."""

from dataclasses import dataclass, field
import logging

logger = logging.getLogger(__name__)


ROLE_PREFIX_MAPPING = {
    "model": "moshi: ",
    "user": "user: ",
}


@dataclass
class TurnManager:
    """Manages VAD state, speaker transitions, text buffering, and context."""

    # VAD parameters
    window_size: int = 4
    threshold: float = 0.5
    warmup_steps: int = 20
    vad_history: list[float] = field(default_factory=list)
    _step_count: int = 0

    # Speaker transition parameters
    stt_wait_steps: int = 0  # Number of steps to wait before switching to model
    _wait_counter: int = 0  # Counter for wait steps
    _pending_speaker: str | None = None  # Track pending speaker switch

    # Conversation/state tracking
    init_active_speaker: str = "model"
    active_speaker: str = "model"
    conversation_context: str = ""
    model_text_buffer: list[tuple[str, str]] = field(default_factory=list)
    user_text_buffer: list[tuple[str, str]] = field(default_factory=list)

    def __post_init__(self):
        # Little trick to make the initial speaker properly displayed
        if self.init_active_speaker == "user":
            self.user_text_buffer = [("user: ", "user")]
        elif self.init_active_speaker == "model":
            self.model_text_buffer = [("moshi: ", "model")]
        self.active_speaker = self.init_active_speaker

    def get_context(self) -> str:
        """Return accumulated conversation context."""
        return self.conversation_context

    def update_vad(self, vad_value: float):
        """Update VAD history with the latest measurement."""
        self._step_count += 1
        self.vad_history.append(vad_value)
        if len(self.vad_history) > self.window_size:
            self.vad_history.pop(0)

    def _handle_pending_speaker_switch(self) -> str | None:
        """Handle counting and pending speaker switch."""
        if self._pending_speaker is None:
            return None
        if self._wait_counter == 0:
            speaker = self._pending_speaker
            self._pending_speaker = None
            return speaker
        self._wait_counter -= 1
        return None

    def _update_active_speaker(self) -> str:
        """Determine new active speaker based on VAD history"""
        current_speaker = self.active_speaker
        if len(self.vad_history) < self.window_size:
            return current_speaker

        vad_neg = all([value > self.threshold for value in self.vad_history])
        vad_pos = sum(self.vad_history) / len(self.vad_history) < self.threshold
        warmup_done = self._step_count >= self.warmup_steps

        if warmup_done and vad_pos and current_speaker == "model":
            if self._pending_speaker != "user":
                self._pending_speaker = "user"
                self._wait_counter = 0
                logger.info(
                    f"[VAD] User started speaking (history={[f'{v:.3f}' for v in self.vad_history]})",
                )
        elif vad_neg and current_speaker == "user":
            if self.model_text_buffer:
                if self._pending_speaker != "model":
                    self._pending_speaker = "model"
                    # After model starts speaking, wait for the specified number of steps before switching speaker
                    # This is to ensure that the stt model has enough time to generate the full user transcript
                    self._wait_counter = self.stt_wait_steps
                    logger.info(
                        f"[VAD] User stopped speaking (history={[f'{v:.3f}' for v in self.vad_history]}), waiting",
                    )
            else:
                logger.info("[VAD] User stopped speaking but LM buffer empty, remaining in user turn")

        switch_speaker = self._handle_pending_speaker_switch()
        if switch_speaker is not None:
            logger.info(f"[VAD] Waiting ended, switching to {switch_speaker}")
            return switch_speaker
        return current_speaker

    def _handle_state_transition(self, new_active_speaker: str) -> list[tuple[str, str]]:
        """Handle state transitions between speakers and flush buffers."""
        outputs: list[tuple[str, str]] = []

        if self.active_speaker != new_active_speaker:
            logger.info(f"[State] Switching to {new_active_speaker}")
            # Add the new active speaker to the outputs and update the active speaker
            outputs.append(("\n" + ROLE_PREFIX_MAPPING.get(new_active_speaker, ""), new_active_speaker))
            self.active_speaker = new_active_speaker

        if new_active_speaker == "user":
            if self.user_text_buffer:
                outputs.extend(self.user_text_buffer)
                buffered_text = "".join([item[0] for item in self.user_text_buffer])
                logger.info(f"[Flushed STT buffer] {buffered_text}")
                self.user_text_buffer = []
        else:
            if self.model_text_buffer:
                outputs.extend(self.model_text_buffer)
                buffered_text = "".join([item[0] for item in self.model_text_buffer])
                logger.info(f"[Flushed LM buffer] {buffered_text}")
                self.model_text_buffer = []

        return outputs

    def handle_spoken_text(self, model_text: str | None = None, user_text: str | None = None) -> list[tuple[str, str]]:
        """Process spoken text for both model and user and decide whether to display or buffer it."""
        new_active_speaker = self._update_active_speaker()
        outputs: list[tuple[str, str]] = self._handle_state_transition(new_active_speaker)

        if model_text is not None:
            if self.active_speaker == "model":
                outputs.append((model_text, "model"))
                logger.info(f"[Display Model] '{model_text.strip()}'")
            else:
                if not self.model_text_buffer:
                    logger.info(
                        f"[Buffer] buffering model text while user is active. first_chunk='{model_text.strip()}'"
                    )
                self.model_text_buffer.append((model_text, "model"))
                logger.info(f"[Buffered Model] '{model_text.strip()}'")
        if user_text is not None:
            if self.active_speaker == "user":
                outputs.append((user_text, "user"))
                logger.info(f"[Display User] '{user_text.strip()}'")
            else:
                if not self.user_text_buffer:
                    logger.info(
                        f"[Buffer] buffering user text while model is active. first_chunk='{user_text.strip()}'"
                    )
                self.user_text_buffer.append((user_text, "user"))
                logger.info(f"[Buffered User] '{user_text.strip()}'")

        self._update_context("".join([item[0] for item in outputs]))

        return outputs

    def _update_context(self, text: str):
        self.conversation_context += text.replace("[RAG]", "")

    def reset(self):
        """Reset VAD state and conversation buffers."""
        self.vad_history = []
        self._step_count = 0
        self._wait_counter = 0
        self._pending_speaker = None
        self.active_speaker = self.init_active_speaker
        self.conversation_context = ""
        self.model_text_buffer = []
        self.user_text_buffer = []
        self.__post_init__()

File: rust/moshi-backend/src/turn_manager.rs
// Copyright (c) Kyutai, all rights reserved.
// This source code is licensed under the license found in the
// LICENSE file in the root directory of this source tree.

use std::collections::VecDeque;

/// Role for text display (model vs user).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TextRole {
    Model,
    User,
}

const ROLE_PREFIX_MODEL: &str = "moshi: ";
const ROLE_PREFIX_USER: &str = "user: ";

/// Number of steps to keep initial speaker before allowing VAD-based switch to the other.
/// Avoids switching to User too early so model greeting text can be shown on the live transcript.
const VAD_WARMUP_STEPS: usize = 12;

/// Manages VAD state, speaker transitions, text buffering, and context for RAG.
#[derive(Debug, Clone)]
pub struct TurnManager {
    pub vad_window_size: usize,
    pub vad_threshold: f32,
    pub vad_wait_steps: usize,

    vad_history: VecDeque<f32>,
    wait_counter: usize,
    pending_speaker: Option<TextRole>,
    pub active_speaker: TextRole,
    pub conversation_context: String,
    model_text_buffer: Vec<(String, TextRole)>,
    user_text_buffer: Vec<(String, TextRole)>,
    step_count: usize,
}

impl TurnManager {
    pub fn new(
        vad_window_size: usize,
        vad_threshold: f32,
        vad_wait_steps: usize,
        init_active_speaker: TextRole,
    ) -> Self {
        let (model_text_buffer, user_text_buffer, active_speaker) = match init_active_speaker {
            TextRole::User => {
                (vec![], vec![(ROLE_PREFIX_USER.to_string(), TextRole::User)], TextRole::User)
            }
            TextRole::Model => {
                (vec![(ROLE_PREFIX_MODEL.to_string(), TextRole::Model)], vec![], TextRole::Model)
            }
        };
        Self {
            vad_window_size,
            vad_threshold,
            vad_wait_steps,
            vad_history: VecDeque::with_capacity(vad_window_size + 1),
            wait_counter: 0,
            pending_speaker: None,
            active_speaker,
            conversation_context: String::new(),
            model_text_buffer,
            user_text_buffer,
            step_count: 0,
        }
    }

    pub fn get_context(&self) -> &str {
        &self.conversation_context
    }

    pub fn update_vad(&mut self, vad_value: f32) {
        self.step_count = self.step_count.saturating_add(1);
        self.wait_counter = self.wait_counter.saturating_sub(1);
        if self.vad_history.len() >= self.vad_window_size {
            self.vad_history.pop_front();
        }
        self.vad_history.push_back(vad_value);
    }

    fn handle_pending_speaker_switch(&mut self) -> Option<TextRole> {
        let pending = self.pending_speaker?;
        if self.wait_counter == 0 {
            self.pending_speaker = None;
            return Some(pending);
        }
        None
    }

    fn update_active_speaker(&mut self) -> bool {
        if self.vad_history.len() < self.vad_window_size {
            return false;
        }
        // Don't switch from Model to User in the first VAD_WARMUP_STEPS steps so initial model text shows on transcript.
        let past_warmup = self.step_count >= VAD_WARMUP_STEPS;
        let vad_avg = self.vad_history.iter().sum::<f32>() / self.vad_history.len() as f32;
        let vad_pos = vad_avg < self.vad_threshold;
        let vad_neg = self.vad_history.iter().all(|&v| v > self.vad_threshold);

        if past_warmup && vad_pos && self.active_speaker == TextRole::Model {
            // Switch to User right away if average vad is positive for over VAD_WINDOW_SIZE steps.
            if self.pending_speaker != Some(TextRole::User) {
                self.pending_speaker = Some(TextRole::User);
                self.wait_counter = 0;
            }
        } else if vad_neg && self.active_speaker == TextRole::User {
            // Switch to Model if
            // 1. vad is negative (user is silent) for consecutive VAD_WINDOW_SIZE steps, and
            // 2. model text buffer is not empty (i.e. model said something during previous user turn or model is currently speaking)
            // after waiting for self.vad_wait_steps steps (to avoid switching during a user PAUSE).
            if !self.model_text_buffer.is_empty() && self.pending_speaker != Some(TextRole::Model) {
                self.pending_speaker = Some(TextRole::Model);
                self.wait_counter = self.vad_wait_steps;
            }
        }

        if let Some(speaker) = self.handle_pending_speaker_switch() {
            if speaker != self.active_speaker {
                self.active_speaker = speaker;
                return true;
            }
        }
        false
    }

    /// Returns list of (text, role) to send to client.
    pub fn handle_spoken_text(
        &mut self,
        model_text: Option<&str>,
        user_text: Option<&str>,
    ) -> Vec<(String, TextRole)> {
        let mut outputs = Vec::new();
        if self.update_active_speaker() {
            let prefix = match self.active_speaker {
                TextRole::Model => format!("\n{ROLE_PREFIX_MODEL}"),
                TextRole::User => format!("\n{ROLE_PREFIX_USER}"),
            };
            outputs.push((prefix, self.active_speaker));
        }

        if self.active_speaker == TextRole::User {
            outputs.append(&mut self.user_text_buffer);
            if let Some(t) = user_text {
                if !t.is_empty() {
                    outputs.push((t.to_string(), TextRole::User));
                }
            }
            if let Some(t) = model_text {
                if !t.is_empty() {
                    self.model_text_buffer.push((t.to_string(), TextRole::Model));
                }
            }
        } else {
            outputs.append(&mut self.model_text_buffer);
            if let Some(t) = model_text {
                outputs.push((t.to_string(), TextRole::Model));
            }
            if let Some(t) = user_text {
                self.user_text_buffer.push((t.to_string(), TextRole::User));
            }
        }

        for (text, _) in &outputs {
            // Do not include [RAG] symbols in the conversation context.
            let cleaned = text.replace("[RAG]", "");
            if !cleaned.is_empty() {
                self.conversation_context.push_str(&cleaned);
            }
        }
        outputs
    }

    #[allow(dead_code)]
    pub fn reset(&mut self, init_active_speaker: TextRole) {
        self.vad_history.clear();
        self.wait_counter = 0;
        self.pending_speaker = None;
        self.active_speaker = init_active_speaker;
        self.conversation_context.clear();
        self.model_text_buffer.clear();
        self.user_text_buffer.clear();
        self.step_count = 0;
        match init_active_speaker {
            TextRole::User => {
                self.user_text_buffer.push((ROLE_PREFIX_USER.to_string(), TextRole::User));
            }
            TextRole::Model => {
                self.model_text_buffer.push((ROLE_PREFIX_MODEL.to_string(), TextRole::Model));
            }
        }
    }
}
